api/v1alpha1/clusterconfig_types.go

```diff
@@ -279,6 +282,28 @@ type User struct {
 	Sudo string `json:"sudo,omitempty"`
 }
 
+// EncryptionAtRest defines the configuration to enable encryption at REST
+// This configuration is used by API server to encrypt data before storing it in ETCD.
+// Currently the encryption only enabled for secrets and configmaps.
+type EncryptionAtRest struct {
+	// Encryption providers
+	// +kubebuilder:default={{aescbc:{}}}
+	// +kubebuilder:maxItems=1
+	// +kubebuilder:validation:Optional
+	Providers []EncryptionProviders `json:"providers"`
+}
+
+type EncryptionProviders struct {
```

Comment:
The apiserver allows only one type of the encryption provider. https://github.com/kubernetes/apiserver/blob/master/pkg/apis/apiserver/validation/validation_encryption.go#L311
`"more than one provider specified in a single element, should split into different list elements"`
`oneof` kubebuilder validation is still not implemented. https://github.com/kubernetes-sigs/controller-tools/issues/461

Should we implement the same validation in the patch handler?

Reply:
We can do, but I also think we can patch the CRD post generation to add the oneOf validation.

Reply:
We can also use cel validation once my upstream pr is finally reviewed and merged...

Reply:
> We can do, but I also think we can patch the CRD post generation to add the oneOf validation.
Thanks. I can try it out. is it okay if I file followup PR for this validation.
> We can also use cel validation once my upstream pr is finally reviewed and merged...
We can remove the workaround once we start using CEL validation is merged

Reply:
I looked into using `oneOf` and unfortunately cluster-api CC variables don't currently support it (see https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/clusterclass_types.go#L424 for what is currently supported). I've asked in [k8s slack](https://kubernetes.slack.com/archives/C8TSNPY4T/p1715774690914879) and will follow up to see if we can add upstream.